### ISSUE TITLE
Discussion on git commit requirements

### ISSUE THREAD

**Issue by cacraigucar** [ESCOMP/CAM#25]
From email  by @gold2718 

I still do not understand the logic of mashing together unrelated commits. The GitHub workflow is smoother and issues are easier to track (and debug) when made separately.
Using this issue and #20 as an example, say for the sake of argument that issue #20 introduced a change (AQP11 to AQPCONST) which is incompatible with my current work. Because there is no separate commit with issue #22 alone, I cannot easily use that while avoiding the #20 change as well.
Merging #20 as a separate PR with a couple of simple tests (already made by @nusbaume) would make no difference in the testing required to merge #20 and make a tag. On the other hand, it makes needed changes available more quickly.
If anyone still disagrees, please add some compelling arguments here.

**Comment by cacraigucar** [comment]
I strongly disagree and believe that EVERY commit should be fully tested and tagged.  If a commit is too inconsequential for complete testing, then it should be merged into the "next tag".   My reasoning is as follows:

1.  By allowing commits to go onto the head of a branch without full testing and tagging, another degree (or more) of freedom has been added to the developer who is the first to do the full testing.  Some commits are hard enough to debug without worrying that the problem is not one that you've introduced.  Every degree of freedom takes developer time to investigate.
2. By introducing the "next tag" label, this is a request to the next developer in line to include these changes in their tag.  If their tag is extremely complicated and runs the risk of being difficult to debug tests, they have the option of not perturbing their testing further.
3. CAM regression testing takes about a day to complete.  If a "minor change" gets pushed to the head of the branch while testing is going on, then testing will need to be restarted.  
4. I do not want to take on discovering bugs that others missed and after work determining that the bug is someone else's.  I'd then need to wait for it to be fixed before I could proceed since its been committed and my tag requires all regression tests to pass.
5. I do not want to push my bugs from an "easy" commit which had hidden consequences on someone else.  Lest you think this won't happen I made a commit to cime after thinking I had thoroughly tested it (but not 100% complete regression testing and documenting).  A couple of days later, I discovered a bug in a single CAM regression test.  Luckily I found this bug before someone else did and pushed out the fix to cime.
6.  As the CAM "integrator" for the last few years, I have seen the instability of untested, yet committed tags.  Whenever I need a tag of cime that has not been brought into an alpha tag, I'd say about half the time, I discover a bug in the head of the cime branch.  It is for that reason whenever possible, I only update externals that have been tested and are in an alpha tag.  I would propose that that we NOT add this same complexity to the CAM branches by allowing "not fully tested" commits onto the CAM trunk.

**Comment by cacraigucar** [comment]
@brian-eaton  replied via email:

We mash together logically distinct change sets into single commits all the
time. Our testing requires enough resources (compute and human) that
separating logically distinct changes into separate commits would be
burdensome. The commit I just made to the release branch contained two
distinct parts; a bug fix and an enhancement. I see no reason to turn that
into 2 commits.

We don't have any guidelines as to what "a couple of simple tests" should
be comprised of. I don't feel we should be burdening the developer who's
making a tested commit deal with finding the problems from a previous
commit that was not adequately tested. I think the lightweight commit
approach that Steve is suggesting works fine in the context of automated
nightly testing and/or a dedicated test engineer. But we currently lack
both of those resources.

I'm not really swayed by the importance of being able to back out a
previous commit that's incompatible with my current branch. It seems to me
that what goes on the trunk is being decided and coordinated taking into
account the goals for the trunk code, and independently of whether a given
trunk commit may conflict with some developer's branch. If the current
trunk is somehow incompatible with my branch, then I have the option to
stay on my branch without updating until I want to get my changes to the
trunk, at which time I need to resolve those conflicts. The resolution
needs to happen sooner or later.

I'm OK with the suggestion Jesse and Cheryl have made in this issue. Using
the issue mechanism does leave a trail for tracking a change, and makes it
easy for the person in line for commit to decide whether or not to pull the
additional presumably small change into their own testing and PR.

**Comment by gold2718** [comment]
> By allowing commits to go onto the head of a branch without full testing and tagging, another degree (or more) of freedom has been added to the developer who is the first to do the full testing. Some commits are hard enough to debug without worrying that the problem is not one that you've introduced. Every degree of freedom takes developer time to investigate.

But this is precisely what the current approach does. In the case of #23, it burdened the tag-making developer (@cacraigucar) with a change made by a different developer (@nusbaume). While this particular case might not be problematic, there have been cases where the contributed code ('can you put this stuff in with your upcoming tag?') causes a problem which the tag-making developer has had to track down. With the 2 commit / 1 tag approach, the tag developer has the option of easily trying out a failing test by leaving out the first commit or trying out that test with the first commit but not the pending tag commit. git provides 21st century tools to make this sort of debugging easier, why not use them?

> I do not want to take on discovering bugs that others missed and after work determining that the bug is someone else's. I'd then need to wait for it to be fixed before I could proceed since its been committed and my tag requires all regression tests to pass.

This is exactly what happens today when someone sends you code to put in your commit.

I agree that all CAM tags should be fully tested. However, if you are really against having the commit be on the trunk then I propose the following (to be required) workflow:

1. The 'unrelated feature developer' (the non-tagger) creates a branch off the head of the trunk and puts their work on it. Let's call this commit `UF-hash`.
2. This commit is added to next tag
3. The tagging developer merges `UF-hash` into their branch before doing pre-tag testing.
4. The tagging developer then goes about the usual testing and tagging process.

This will preserve a separate commit for unrelated features without making a separate trunk commit. It is merely a first parent vs second parent subtlety but it might increase your comfort level.

Thoughts?

**Comment by cacraigucar** [comment]
Jesse and I can try this procedure with the release branch of #23 and #20

**Comment by brian-eaton** [comment]
I think this proposed workflow which puts untested commits on branches rather than the "trunk" seems like it would work.  It's big advantage over putting the untested commits on the trunk is that the developer who's making the next tagged commit doesn't have to deal with it if it would be an undue burden.  And that's the same advantage that our current workflow provides over the originally proposed "2 commit / 1 tag" approach.

**Comment by cacraigucar** [comment]
@gold2718 Can this issue be closed?

**Comment by gold2718** [comment]
I suppose, although I still do not understand why folks are worried about my proposal when there are many untested and untagged commits going into cam_development (and the maintenance branches) with every tag commit!
In the SVN repo, we tried to only commit to the trunk branch when the code was ready to be copied into a tag.
Here, everyone is arguing against having untested commits go into cam_development *and yet, that is exactly what everyone is doing*. This could be avoided using rebase but few folks seem to want to do that either.

So, no one likes my proposal but no one is also being careful to only add tested commits. Feel free to close the issue if you are okay with that status.
